### ISSUE TITLE
WebSocket connection dies

### ISSUE THREAD

**Issue by yangmillstheory** [layerhq/layer-websdk#61]
This is using Chrome Version 50.0.2661.102 (64-bit) and Firefox 46.0.1.

In the sample React app, when I authenticate using the staging service on Heroku, after some time the WebSocket connection dies and doesn't come back (even after a page refresh).
I've seen the same behavior in my own integration, using Auth0 is the authentication broker. Here are some examples due to the session expiring

![jun-05-2016 15-49-54](https://cloud.githubusercontent.com/assets/2729079/15808772/2f0e3d88-2b35-11e6-9b18-dc74f4e33bb0.gif)

![image](https://cloud.githubusercontent.com/assets/2729079/15808777/440eff4c-2b35-11e6-95e4-27288fe436d2.png)

Here's an example where the socket just dies, for no apparent reason

![image](https://cloud.githubusercontent.com/assets/2729079/15812941/6ef72178-2b6f-11e6-96cb-0311a019b778.png)

This also happens in an Incognito window in Chrome.

Should we be doing anything to reauthenticate (issue `challenge` again), or does this happen transparently after some time (doesn't seem like it)? The Web SDK docs seem to suggest leaving this alone

![image](https://cloud.githubusercontent.com/assets/2729079/15808139/9ec434a4-2b24-11e6-804a-553dccbde0bf.png)

but in my case nothing happens until I refresh the page and receive more messages from other parties. (Note that I can't actually send a message because the web socket is dead.)

I've I leave the app idle, there can be hundreds of failed WebSocket connection messages in the console, so it leads me to think it's a pretty bad state an app should actively be trying to avoid.

When the connection dies due to the session expiring, should the Layer client automatically try to refresh the session?

**Comment by yangmillstheory** [comment]
It looks like #58, #59 would solve these problems - hope those get merged soon!

**Comment by michaelkantor** [comment]
Going to try to get that shipped today.  Yes, that's exactly what #58 is there for, but more importantly, timing out staging apps after a short period is a new change on the server to insure apps are properly handling this event, whereas production apps would do this every 30 days rather than every few minutes.

**Comment by yangmillstheory** [comment]
Thanks @michaelkantor. 

Is there a recommended/example way for an app to handle this event? I was under the impression [from the README](https://github.com/layerhq/layer-websdk/blame/master/README.md#L505) that socket management was up to the library (mostly).

**Comment by michaelkantor** [comment]
Socket Management is handled by the Web SDK.

The Client API has recently started expiring Authentication after a short interval (5 minutes I think) for Staging apps to insure developers of stating apps are prepared to handle authorization expiration which typically happens after 30 days (on production apps).

Handling of this is done entirely by your `challenge` event handler which must get an Identity token from your server. This is not Socket Management, this is Authentication which falls in your court.

**Comment by yangmillstheory** [comment]
OK. So once #58 is merged, will any sockets dropped due to an expired session cause the `challenge` to be re-emitted? 

If so, since the `challenge` handler is defined properly in the app (even [here](https://github.com/layerhq/layer-js-sampleapps/blob/master/common/layerSample.js#L26)), can we expect the websocket to recover transparently?

**Comment by michaelkantor** [comment]
Thats exactly right.  Merge coming soon.

**Comment by yangmillstheory** [comment]
Thanks @michaelkantor!

**Comment by michaelkantor** [comment]
WebSDK 1.0.7 is released on NPM, CDN and Github.

**Comment by yangmillstheory** [comment]
Terrific, this looks like it worked. Notice the pattern of requests.

![image](https://cloud.githubusercontent.com/assets/2729079/15844020/9c1defd6-2c1b-11e6-8b9a-893d78bd2b3c.png)

The first session was at 

> Date:Tue, 07 Jun 2016 02:05:00 GMT

After a bit, it expired at 

> Date:Tue, 07 Jun 2016 02:15:26 GMT

which immediately triggered

> Date:Tue, 07 Jun 2016 02:15:27 GMT

It looks like the time to expiry is currently ~10 minutes.

When we make the request to the Layer root URI it looks like it returns a nonce

![image](https://cloud.githubusercontent.com/assets/2729079/15844058/023945ae-2c1c-11e6-8181-6f14eaf6f009.png)

and it looks like the code transparently answers the challenge.

**Comment by jcyusupov** [comment]
@michaelkantor Hi! I have the same issue now for version 1.0.9. Could you please give me any advise?

**Comment by michaelkantor** [comment]
Is your `challenge` event not being triggered when your session expires?
Do you use `on('challenge'...)` or `once('challenge'...)`?  If `once` change to `on` so it can be retriggered.

**Comment by yangmillstheory** [comment]
@michaelkantor this is still an issue; here's what's happening with my staging app:

- get challenge with nonce, send to my server
- my server signs JWT, posts to /sessions successfully
![image](https://cloud.githubusercontent.com/assets/2729079/23290609/f0358886-fa06-11e6-9bd9-06a7242e65d9.png)
- ~30 seconds later, there's a call to 
![image](https://cloud.githubusercontent.com/assets/2729079/23290616/f98f5650-fa06-11e6-8690-86e016583929.png)
which fails

I have the same code path working for the production app. Why would a call to /sessions succeed then the call to https://api.layer.com/?client=websdk1.0.13 fail?

Note that that the latter endpoint also returns:

![image](https://cloud.githubusercontent.com/assets/2729079/23290637/2589f954-fa07-11e6-9bb9-5ce779674e7a.png)

but I'm not sure what to do with this.

Please advise?

**Comment by yangmillstheory** [comment]
It looks like those failures are due to this code path https://github.com/layerhq/layer-websdk/blob/8a5a66cbed4166407226a7e139c473dea66a7c46/src/websockets/socket-manager.js#L572 in the 1.x release.

**Comment by michaelkantor** [comment]
Do you get the same results if you update to the latest WebSDK? Looks like your still running `1.0.13`.  

Do you have a url + authentication to a non-minified version of this app showing this error? (if so, you can gmail me via github account's email address).

**Comment by yangmillstheory** [comment]
Hi @michaelkantor,

I'm going to try upgrading the latest SDK in a separate task; it's a pretty invasive change for us. The app is currently private, so I don't think it would be easy to share the actual URL. I might be able to share a basic GitHub demo.

Is this a known bug that never was fixed in the 1.x line?
